docs(firestore): added pagination using firestore

yslDevelop opened the pull request:
### Description

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request properly. -->
<!-- Explain the **motivation** for making this change e.g. what existing problem does the pull request solve? -->

I thorught I could tackle down this problem since I spent so much time on figuring out how pagination actually works. 

### Related issues

<!-- If this PR fixes an issue, include "Fixes #issueNumber" to automatically close the issue when the PR is merged. -->

Fixes #6055 

### Release Summary

<!-- An optional description that you want to appear on the generated changelog -->

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No


### Test Plan

<!-- Demonstrate the code you've added is solid, e.g. test logs or screenshots. -->
![second page load](https://user-images.githubusercontent.com/93795660/160174306-38980862-deb5-44c6-8997-08d3dda8b0bb.png)
![before load](https://user-images.githubusercontent.com/93795660/160174307-33375351-e735-4ca2-a222-45f6ff471512.png)
![first page load](https://user-images.githubusercontent.com/93795660/160174308-4796a112-a1af-4088-82b6-8d975ae0eb48.png)
---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

:fire: :fire: :fire: :fire: :fire: :fire:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/react-native-firebase?pullRequest=6157) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/react-native-firebase?pullRequest=6157) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/react-native-firebase?pullRequest=6157) it.</sub>

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+---
+title: Pagination
+description: Pagination example using Cloud Firestore.
+next: /functions/usage
+previous: /firestore/usage-with-flatlists
```
In order for this to work, those two pages need to have changes as well to "splice" this page in, since they each have next/prev links.
I think /functions/usage will need it's prev link altered to no longer be /firestore/usage-with-flatlists (it should be /firestore/pagination now)
I think /firestore/usage/with-flatlists will need it's next link changed to firestore/pagination instead of /functions/usage

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+previous: /firestore/usage-with-flatlists
+---
+
+Pagination using cloud firestore can be done in various ways but here's the very basic way to do it.
+We are going to make the pagination using querying features provided by firestore.
```
```suggestion
Pagination using cloud firestore may be done in various ways but here's a basic way to do it using the firestore query features:
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+
+# Setup state
+
+First, setup a component which will display the list of data.
+The component will have 2 separate states; `last` and `userData`:
```
A bit more compact but most importantly the variable is most correctly named lastDocument in the code so reference it as such

```suggestion
First, create a list display component with 2 state items; `lastDocument` and `userData`:
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+
+# `LoadData` function
+
+Next, we'll make a function called `LoadData` that fetches data from `Users` collection.
+This function will be hooked with the `Button`.
```
```suggestion
Next, make a function called `LoadData` that fetches data from `Users` collection, and call it when a `Button` is pressed.
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+const userCollection = firestore().collection('Users');
+
+const App: () => Node = () => {
+  const [lastSnapshot, setLastSnapshot] = useState(null);
```
```suggestion
  const [lastDocument, setLastDocument] = useState();
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+Next, we'll make a function called `LoadData` that fetches data from `Users` collection.
+This function will be hooked with the `Button`.
+
+If lastSanpshot is not assigned (meaning initial load), the function will fetch from the start.
+After successful fetch from the collection, store the last snapshot data by `setLastSnapshot`.
```
```suggestion
If lastDocument is undefined the function will fetch from the star to perform the initial loadt, otherwise it use the lastDocument from the snapshot
After successful fetch from the collection, store the last document we accessed with `setLastDocument`.
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+const userCollection = firestore().collection('Users');
+
+const App: () => Node = () => {
+  const [lastSnapshot, setLastSnapshot] = useState(null);
```
```suggestion
  const [lastDocument, setLastDocument] = useState();
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+
+  function LoadData() {
+    console.log('LOAD');
+    if (!lastSnapshot) {
+      userCollection
+        .orderBy('age') //sort the data
+        .limit(3) //get limited amount of data
+        .get()
+        .then(querySnapshot => {
+          setLastSnapshot(querySnapshot.docs[querySnapshot.docs.length - 1]); //set up last snapshot for pagination
+          MakeUserData(querySnapshot.docs); //do what you need to do with the data
+        });
+    } else {
+      userCollection
+        .orderBy('age')
+        .startAfter(lastSnapshot) //fetch data that is placed after the last snapshot that we fetched before.
+        .limit(3)
+        .get()
+        .then(querySnapshot => {
+          setLastSnapshot(querySnapshot.docs[querySnapshot.docs.length - 1]);
+          MakeUserData(querySnapshot.docs);
+        });
+    }
```
I think this de-duplicates the code?

```suggestion
    let query = userCollection.orderBy('age'); // sort the data
    if (lastSnapshot !== undefined) {
      query = query..startAfter(lastSnapshot); // fetch data following the last document accessed
    }
    query.limit(3) // limit to your page size, 3 is just an example
        .get()
        .then(querySnapshot => {
          setLastDocument(querySnapshot.docs[querySnapshot.docs.length - 1]);
          MakeUserData(querySnapshot.docs);
        });
    }
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+
+# `MakeUserData` function
+
+This is just an example function. You may change this functino into things that you need.
```
```suggestion
This is just an example function, alter it to process the data to meet your requirements.
```

mikehardy commented on docs/firestore/pagination.md:
````diff
@@ -0,0 +1,187 @@
+}
+```
+
+Now, everytime you click the button, you will load the data from the `Users` collection.
+If the fetch succesfully completed, you will save the last snapshot and set the userdata component.
+Based on the last snapshot, `LoadData` will fetch the data accordingly.
````
```suggestion
Now, every time the button is pressed, `Users` collection data will be fetched one page at a time.
```

mikehardy commented on docs/firestore/pagination.md:
```diff
@@ -0,0 +1,187 @@
+const userCollection = firestore().collection('Users');
+
+const App: () => Node = () => {
+  const [lastSnapshot, setLastSnapshot] = useState(null);
+  const [userData, setUserData] = useState([]);
+
+  function LoadData() {
```
If you like the changes above to call it "lastDocument" and to de-duplicate the load code, then the same should be pasted here but I'll wait to see what you think

mikehardy reviewed (changes requested):
Apologies for the long delay here @yslDevelop ! I just got a batch of time to focus on firebase tasks and saw this needed review. Gave it a scan and I really like it. I proposed a bunch of changes but I recognize it is easy to be a critic, and creating the thing in the first place is the real difficulty. My edits are just small grammar items and then an attempt to de-duplicate the code really.

What do you think?

yslDevelop commented:
Thank you for the thorough feedback. I've changed everything I could find that you wanted me to change :) also, I agree that lastDocument is more intuitive so I have changed lastSnapshot to lastDocument! Please tell me if there's anything I should do more. Thank you!

mikehardy commented:
>  I've changed everything I could find that you wanted me to change :)

Thanks! I think this is close, but -

1-  now there is a new / second file in the diff view called `basic pagination.md` that should not be there I think?
2- the next/prev links on the *other* pages are still not pointing at this page - to add a new page to the document set requires changing three files - a) adding the new file with next/prev links to the files before and after your added page b) changing the file that is the page linked as "next" in your new file so that it's "prev" is now your new file and c) changing the file that is the page linked as "prev" in your new file so that it's "next" link is now your new file. That makes the next/prev links work before your page, on your page, and on the next page
3- the code for initial load versus follow-on loads is de-duplicated now, but the "complete example" code listing still has the old style with duplication - the de-duplicated style needs to be pasted in there I think?

yslDevelop commented:
Thank you for the fast feedback!
I'm sorry for the confusion. basic-pagination was a exercise for me :( I have deleted it.
Also, I have changed the usage-with-flatlist's next and funciton/usage/index's previous
Lastly, I fixed the de-duplcation issue too :)

mikehardy reviewed (dismissed):
Excellent! This looks really great - you can see it on the Vercel test deployments https://react-native-firebase-pgtakgkqk-invertase.vercel.app/firestore/pagination

When I went to check it (pseudo-)live I noticed it wasn't in the sidebar, and realized it's easier to just do it on your fork/branch then explain it, so I added one last commit.

If the Vercel deployment looks good then for navigation (the page itself already looks good...) then this is ready to go!

mikehardy commented:
Also added words to spellcheck and yarn prettier on it and pushed those - just cramming this through CI :-)
